partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeEnv.kt

```diff
@@ -19,7 +20,10 @@ import org.partiql.value.stringValue
 /**
  * TypeEnv represents a variables type environment.
  */
-internal class TypeEnv(public val schema: List<Rel.Binding>) {
+internal data class TypeEnv(
+    public val schema: List<Rel.Binding>,
+    public val stack: List<TypeEnv>
```

Comment:
```
s/stack/scopes
s/height/depth
```

Reply:
In the section you referenced in crafting interpreters, the depth is the difference between the current scope and the variable being referenced. I'm moving away from that because we don't need to know the difference. We can just index the stack as-is.